src/repo/remote.rs

```diff
@@ -0,0 +1,692 @@
+                index_checksum,
+                pack_checksum,
+            });
+        }
+
+        Ok(Self {
+            path: None,
+            meta,
+            url,
+            packs,
+        })
+    }
+
+    fn read_keyed_line<R: BufRead>(lines: &mut io::Lines<R>, key: &str) -> Result<String, Error> {
+        let line = match lines.next() {
+            None => Err(Error::BadRemoteIndexFormat(
+                format!("Expected '{} ...', but end of file was reached!", key),
+                None,
+            )),
+            Some(Err(e)) => Err(Error::IOError(e)),
+            Some(Ok(line)) => Ok(line),
+        }?;
+
+        let mut parts = line
+            .split(Self::is_field_separator)
+            .filter(|&p| !p.is_empty());
+
+        if parts.next() != Some(key) {
+            return Err(Error::BadRemoteIndexFormat(
+                format!("Expected '{} ...': '{}'", key, line),
+                None,
+            ));
+        }
+
+        let value = parts.next().ok_or_else(|| {
+            Error::BadRemoteIndexFormat(
+                format!("Expected a value after '{}': '{}'", key, line),
+                None,
+            )
+        })?;
+
+        if parts.next().is_some() {
+            return Err(Error::BadRemoteIndexFormat(
+                format!("Too many values: '{}'", line),
+                None,
+            ));
+        }
+
+        Ok(value.to_string())
+    }
+
+    fn is_field_separator(ch: char) -> bool {
+        ch == ' ' || ch == '\t'
+    }
+}
+
+impl std::fmt::Display for RemoteIndex {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+        if let Some(stem) = self.path().and_then(|p| p.file_stem()) {
+            write!(fmt, "{} ({})", stem.to_string_lossy(), self.url)?;
+        } else {
+            let stem = self.url.rsplit_once('/').unwrap().0;
+            write!(fmt, "{} ({})", stem, self.url)?;
+        }
+        Ok(())
+    }
+}
+
+/// Loads all remote index (.esi) files from the target directory.
+pub fn load_remotes(base_path: &Path) -> Result<Vec<RemoteIndex>, Error> {
+    let paths = fs::read_dir(base_path).map_err(Error::IOError)?;
+    paths
+        .filter_map(|e| e.ok())
+        .filter(|p| p.path().extension() == Some(OsStr::new(REMOTE_INDEX_EXTENSION)))
+        .map(|p| RemoteIndex::load(p.path()).reify(p.path().display()))
+        .collect()
+}
+
+/// Sends an HTTP GET request with the specified URL.
+///
+/// # Arguments
+/// * `url` - the URL to fetch
+/// * `timeout` - the timeout for the whole of the request
+/// * `size_limit` - the maximum size of the response body, before the
+///   connection is terminated
+/// * `if_modified_since` - Sets the value of the `If-Modified-Since` HTTP
+///   header
+fn open_remote_resource(
+    url: &Url,
+    timeout: Option<Duration>,
+    if_modified_since: Option<SystemTime>,
+) -> Result<Option<(usize, impl Read)>, Error> {
+    let mut agent_builder = AgentBuilder::new();
```

Comment:
I think the agent should be created outside this function, architecturally. It may hold connection pools and whatnot (but I'm speculating here, and I did the experiment in the past and it showed no performance benefit, that I erroneously expected).

Reply:
I will try to refactor this so it can be passed from higher up the call stack.